pyfstat/injection_parameters.py

```diff
@@ -1,83 +1,256 @@
+    """
+
+    function_name = f"{prior_function.__name__}"
+    if function_name in _pyfstat_custom_priors:
+        raise ValueError(
+            f"Custom prior `{function_name}` already defined in "
+            "`pyfstat._pyfstat_custom_priors.` "
+            "Please, use a different function name"
+        )
+
+    _pyfstat_custom_priors[function_name] = prior_function
+
+    return prior_function
+
+
 isotropic_amplitude_priors = {
-    "cosi": {"uniform": {"low": -1.0, "high": 1.0}},
-    "psi": {"uniform": {"low": -0.25 * np.pi, "high": 0.25 * np.pi}},
-    "phi": {"uniform": {"low": 0, "high": 2 * np.pi}},
+    "cosi": {"stats.uniform": {"loc": -1.0, "scale": 2.0}},
+    "psi": {"stats.uniform": {"loc": -0.25 * np.pi, "scale": 0.5 * np.pi}},
+    "phi": {"stats.uniform": {"loc": 0, "scale": 2 * np.pi}},
 }
 
 
-@define(kw_only=True, slots=False)
+@custom_prior
+def uniform_sky_declination(generator: np.random.Generator) -> float:
+    """
+    Return declination values such that, when paired with right ascension
+    values sampled uniformly along [0, 2*pi], the resulting pairs of samples
+    are uniformly distributed on the 2-sphere.
+
+    Parameters
+    ----------
+    generator:
+        As required by InjectionParametersGenerator.
+
+    Returns
+    -------
+    declination:
+        Declination value distributed
+    """
+    return np.arcsin(generator.uniform(low=-1.0, high=1.0))
+
+
 class InjectionParametersGenerator:
     """
     Draw injection parameter samples from priors and return in dictionary format.
 
     Parameters
     ----------
-    priors: dict
+    priors:
         Each key refers to one of the signal's parameters
         (following the PyFstat convention).
-        Priors can be given as values in three formats
-        (by order of evaluation):
 
-        1. Callable without required arguments:
-        `{"ParameterA": np.random.uniform}`.
+        Each parameter's prior should be given as a dictionary entry as follows:
+        `{"parameter": {"<function>": {**kwargs}}}`
+        where <function> may be (exclusively) either a user-defined function decorated
+        with `@custom_prior` or the name of a `scipy.stats` random variable.
+
+        * | If a user-defined function is used, such a function *must* take
+          | a `generator` kwarg as one of its arguments and use such a generator
+          | to generate any required random number within the function.
+          | The `generator` kwarg is *required* regardless of whether this is a
+          | deterministic or random function.
+          | For example, a negative log-distributed random number could be constructed as
+
+        ::
 
-        2. Dict containing numpy.random distribution as key and kwargs in a dict as value:
-        `{"ParameterA": {"uniform": {"low": 0, "high":1}}}`.
+            @pyfstat.custom_prior
+            def negative_log_uniform(generator):
+                return -10**(generator.uniform())
```

Comment:
should have an explicit example of how to instantiate an object of this class with that same custom prior (best in that same example code box, going all the way to draw()